homeassistant/components/recorder/purge.py

```diff
@@ -86,52 +88,80 @@ def purge_old_data(
+        if event.state_id:
+            state_ids.add(event.state_id)
+        if event.attributes_id:
+            attributes_ids.add(event.attributes_id)
+    return event_ids, state_ids, attributes_ids
 
 
 def _select_unused_attributes_ids(
-    session: Session, attributes_ids: set[int]
+    session: Session, attributes_ids: set[int], using_sqlite: bool
 ) -> set[int]:
     """Return a set of attributes ids that are not used by any states in the database."""
     if not attributes_ids:
         return set()
-    to_remove = attributes_ids - {
-        state[0]
-        for state in session.query(distinct(States.attributes_id))
-        .filter(States.attributes_id.in_(attributes_ids))
-        .all()
-    }
+
+    if using_sqlite:
+        id_query = session.query(distinct(States.attributes_id)).filter(
+            States.attributes_id.in_(attributes_ids)
+        )
+    else:
+        #
+        # We are jumping through hoops a bit here to handle
+        # a case where sqlite performance on finding unused attributes_id
+        # far exceeds other DBMS
+        #
+        # MariaDB/MySQL cannot optimize that query well and has to examine
+        # all the rows that match
+        #
+        # > explain select distinct attributes_id from states where attributes_id in (136723);
+        # +------+-------------+--------+------+-------------------------+-------------------------+---------+-------+-------+-------------+
+        # | id   | select_type | table  | type | possible_keys           | key                     | key_len | ref   | rows  | Extra       |
+        # +------+-------------+--------+------+-------------------------+-------------------------+---------+-------+-------+-------------+
+        # |    1 | SIMPLE      | states | ref  | ix_states_attributes_id | ix_states_attributes_id | 5       | const | 22842 | Using index |
+        # +------+-------------+--------+------+-------------------------+-------------------------+---------+-------+-------+-------------+
+        #
+        # With a single query, it can be optimized away
+        #
+        # > explain select min(attributes_id) from states where attributes_id = 136723;
+        # +------+-------------+-------+------+---------------+------+---------+------+------+------------------------------+
+        # | id   | select_type | table | type | possible_keys | key  | key_len | ref  | rows | Extra                        |
+        # +------+-------------+-------+------+---------------+------+---------+------+------+------------------------------+
+        # |    1 | SIMPLE      | NULL  | NULL | NULL          | NULL | NULL    | NULL | NULL | Select tables optimized away |
+        # +------+-------------+-------+------+---------------+------+---------+------+------+------------------------------+
+        #
+        # We test this path to make sure it will work with
+        # sqlite, however only on tests that have < 500 ids
+        # since otherwise we hit the limit
```

Comment:
Please try to rephrase this comment:
- Where and how is this path tested?
- Why do we need to ensure it works with sqlite?
- What limit do we hit?

Reply:
Adjusted to explain that we don't have postgresql or mysql running in tests, and we currently test this path using sqlite as long as we do not hit the 500 union limit in sqlite.